url46a88cc932_toby1984_tinyc_tgz-pJ8-de_codesourcery_dcpu16_compiler_CompilerJ8/src/de/codesourcery/dcpu16/compiler/DataType.java

```diff
@@ -0,0 +1,314 @@
+    {
+    	final Map<Identifier,Map<Integer,DataType> > cache;
+    	if ( isArray ) {
+    		cache=ARRAY_INSTANCE_CACHE;
+    	} else {
+    		cache = INSTANCE_CACHE;
+    	}
+    	
+    	Map<Integer, DataType> map = cache.get( identifier );
+    	if ( map == null ) {
+    		map = new HashMap<>();
+    		cache.put( identifier , map );
+    	}
+    	DataType result = map.get( ptrCount );
+    	if ( result != null ) {
+    		return result;
+    	}
+    	result = new DataType(identifier,isArray,isBuiltInType(identifier),ptrCount);
+    	map.put(ptrCount,result);
+    	return result;
+    }
+    
+    private static void putDataType(DataType type) 
+    {
+    	final Map<Identifier,Map<Integer,DataType> > cache;
+    	if ( type.isArray ) {
+    		cache=ARRAY_INSTANCE_CACHE;
+    	} else {
+    		cache = INSTANCE_CACHE;
+    	}
+    	
+    	Map<Integer, DataType> map = cache.get( type.identifier );
+    	if ( map == null ) {
+    		map = new HashMap<>();
+    		cache.put( type.identifier , map );
+    	}
+    	
+    	map.put(type.ptrCount,type);
+    }
+    
+    public static boolean isBuiltInType(Identifier identifier) 
+    {
+    	switch( identifier.getStringValue() ) {
+    		case "void":
+    		case "boolean":
+    		case "char":
+    		case "int":
+    			return true;
+    	}
+    	return false;
+    }
+    
+    protected Object internalExtractValue(ASTNode node) 
+    {
+        throw new UnsupportedOperationException("internalExtractValue() not implemented for data type "+this);        
+    }
+    
+    public final Object extractValue(ASTNode node) 
+    {
+        Object result = internalExtractValue(node);
+        if ( result == null ) {
+            throw new IllegalArgumentException("Data type "+this+" cannot extract a value from "+node);
+        }
+        return result;
+    }
+    
+    public boolean isArray()
+    {
+        return isArray;
+    }
+    
+    public boolean isIntegerType() 
+    {
+    	return isBasicType() && isAssignableTo( DataType.INTEGER ); 
+    }
+
+    public boolean isBuiltIn() {
+        return isBuiltIn;
+    }
+    
+    public Identifier getIdentifier()
+    {
+        return identifier;
+    }
+    
+    @Override
+    public String toString()
+    {
+    	String result = identifier.getStringValue();
+    	for ( int i = 0 ; i < ptrCount ; i++) {
+    		result = result+"*";
+    	}
+    	if ( isArray ) {
+    		result+="[]";
+    	}
+        return result;
+    }
+
+    @Override
+    public int hashCode()
+    {
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + identifier.hashCode();
+        result = prime * result + (isArray ? 1231 : 1237);
+        result = prime * result + (isBuiltIn ? 1231 : 1237);
+        return result;
+    }
+    
+    public final boolean isAssignableTo(DataType other) {
+        return other.isAssignableFrom( this );
+    }
+    
+    public final boolean isAssignableFrom(DataType other) 
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.